AI Tutor: Add custom prompt via S3

edcodedotorg opened the pull request:
This PR adds the ability for PMs and curriculum team members to create a a custom prompt for AI Tutor and share it with students and educators via a URL (e.g. for A/B tests during pilot). 

Prompts are created in Amazon S3 in the `cdo-curriculum/media/prompt-library` bucket.  Keep the filenames lowercase without spaces or special characters (dash `-` is ok) and give them a `.md` extension:

<img width="1119" height="665" alt="Screenshot 2025-09-02 at 10 04 19 AM" src="https://github.com/user-attachments/assets/7eebd281-4433-4a70-881c-b9b210602cfa" />

Once you have added a prompt you can share it with users for testing (not for production) by setting the URL's `aitutor-custom-prompt` query param to the file name minus the `.md` (`?aitutor-custom-prompt=name-without-dot-md`).  

Note that an incorrect name (spaces or special characters or no `.md`) will not load so you can look in the console for a `🤖: systemPrompt:` message to check which prompt was loaded.

Here is an example of tutor running with a custom "pirate" prompt (let's delete once reviewed):

<img width="1040" height="763" alt="Screenshot 2025-08-26 at 3 18 43 PM" src="https://github.com/user-attachments/assets/4b2c0ce9-b16c-4027-a76f-9874a020b102" />

<br></br>
If no query param is given or if loading fails due to incorrect filename or a timeout, the default prompt will load:

<img width="1042" height="755" alt="Screenshot 2025-08-26 at 3 15 45 PM" src="https://github.com/user-attachments/assets/5a5f1991-fd61-4101-838e-28160aba2964" />

## Implementation details

This work initially used a Github "gist" to store and load the prompt but we felt we should limit the ability to create a shareable prompt to our org. It then used a separate GitHub repo to ensure that only members of our org (or added guests) could create sharable prompts. This also brought the GitHub web editor UI and version history for free.  However there were various concerns that this use of GitHub as a runtime dependency (even in a limited, testing setting) created a bad precedent so the storage was moved to S3 without "all the bells and whistles" (including version control) of the GitHub/Gist solutions.

## Follow-up work

Explore a more long term solution that ideally would include the version control and collaboration of git, fast updates, and a more modular approach to prompt engineering.

breville commented on apps/src/lib/util/customPrompt.ts:
```diff
@@ -0,0 +1,43 @@
+ *    });
+ **/
+
+export const fetchCustomPrompt = async (
+  promptName: string,
+  timeoutMs = 10000
+) => {
+  const url = `https://raw.githubusercontent.com/code-dot-org/experimentation-settings/refs/heads/main/tools/aitutor/${promptName}.md`;
+  const controller = new AbortController();
+
+  //Timeout and use default system prompt.
+  const timeout = setTimeout(() => controller.abort(), timeoutMs);
+
+  try {
+    const response = await fetch(url, {signal: controller.signal});
+
+    if (!response.ok) {
+      throw new Error(`Fetch error: ${response.status}`);
+    }
+
+    const text = await response.text();
+    return text;
+  } catch (err) {
+    return null;
+  } finally {
+    clearTimeout(timeout);
+  }
+};
```
I think we could use our in-house `HttpClient.get`?

edcodedotorg replied:
Hmm.  It [doesn't seem to support a timeout](https://github.com/code-dot-org/code-dot-org/blob/1ed446f853c9d70c79037700b8f3fb54a86e0e9c/apps/src/util/HttpClient.ts#L115). I could add a timeout but then the caller has to populate two optional parameters (destructuring an options object would probably be better next time):

```
async function get(
  endpoint: string,
  useAuthenticityToken = false,
  headers: Record<string, string> = {},
  timeoutMs = 1000
)
```

Ok to leave it?  Fetch is what's used under the hood in `HttpClient.get`

breville commented on apps/src/lab2/views/components/AiTutor2Chat.tsx:
```diff
@@ -100,6 +108,31 @@ const AiTutor2Chat: React.FunctionComponent<AiTutor2ChatProps> = ({
+
+  useEffect(() => {
+    if (customPromptName) {
+      fetchCustomPrompt(customPromptName).then(prompt => {
```
What happens if the user starts using chat before the custom prompt has been set, say when there is a slow network?

edcodedotorg replied:
[ChatWorkspace won't load](https://github.com/code-dot-org/code-dot-org/blob/5c238b9e5ed12b0b46c67b5849acb32063201c6f/apps/src/lab2/views/components/AiTutor2Chat.tsx#L146) until the `systemPrompt` state variable is set which happens as soon as the response is received (will use default prompt if response not successful) or in the worst case it will load with the default prompt after a 10 second timeout (if the request hangs).

edcodedotorg replied:
Also, I added a spinner when loading the custom prompt:

<img width="1114" height="754" alt="Screenshot 2025-08-28 at 11 28 12 AM" src="https://github.com/user-attachments/assets/716f01ef-2353-4510-85bf-979b3f5c35e9" />

edcodedotorg commented:
@snickell all great points.  I think the issue I have with just whipping up something that "fits" into our normal "put it in a DB behind an admin interface" MO, is that this doesn't really fit that model.  I think the hack is **_serving_** the prompts from GitHub, **_not necessarily_** storing the prompts in GitHub.

This is all just a hunch but I believe these prompts **_want_** to be in version control where they can be collaborated on widely, versioned, and possibly ingested and transformed easily with local developer tools.  I feel the problem might not be "using github as a DB" but rather if we jump to using a DB, where a git versioned filesystem could actually be the best abstraction.

I know we do this for level data but this seems different for two reasons:

1. The prompt files are human readable and editable in the format they would be stored in the filesystem/git
2. In certain situations, changes to the prompts should and can be made available immediately and do not need to be tested and deployed in lock-step with the rest of our codebase the way level data does (this isn't to say there will not need to be testing but just that it is different than how level data needs to be tested)

Besides the idea that this "bad hack" could be around in a year, I think the worst case scenario with the interim approach is that a test prompt can fail to load (for any reason including a school blocking GitHub, an unexpected rate limit, or a GitHub outage) causing the tutor to revert to the default prompt.  The worst case scenario with engineering a "real" solution, involving directly storing to a DB, is that we miss the opportunity to evaluate all the affordances of storing these prompts in normal files in a GitHub repo.

> Questions:
> 
> 1. Is it potentially possible to do a related but different hack which doesn't pull in this complexity where you actually encode the custom prompt as a URI query parameter? Yes these would be very long links to send out, but for whatever reason that's a hack that does not bother me.

Yeah we discussed that and we could even pass it through a URL shortener.  But again I'm not just trying to offer an interim solution but am trying to learn what the longer term solution wants to be.

> 2. Are the experiments going to be handing out links to prod? If so, can we just add a simple rails controller and model to store these prompts in the main SQL DB? Could we limit access to this controller to folks that logged in using @code.org SSO (I wouldn't be surprised is this was either very easy or very hard depending on what exists today)?

Yes, and I looked into some routes/controllers for inspiration but ended up feeling it was wrong to try to do it the "right" way without at least a little exploration and user testing (with our team I mean).  If I had to guess and make a decision today, I'd say I would advocate for storing the prompts in Git and then I'd be looking at how levelbuilder ingests the git repo and populates the DB.  But I know this might be just as controversial so I'd like to get some more data first.


> **If ya'll do decide to go forward with this approach because you judge the hack is worth the downsides**: I would want to see neon flashing warning comments all over explaining that this is a very bad hack, giving conditions for how its used ("only to be used during first wave of iterating ai tutor prompts with curriculum"), and finally giving a "drop dead" date when anyone encountering it can remove it from the codebase without understanding where it is or where the project is in its development state. With that in mind, I also don't like seeing the hack broken out into its own src/lib/customPrompt.js file, I'd rather embed the hack inside the JSX itself isolating the damage close to usage.

Sure. Are you ok with a drop dead date of January 1, 2026?  And yeah I had originally put it all in the `.jsx` file (in a previous PR I reverted this in) but I moved it more so to define the interface with an eye on swapping out the implementation later.  But I can pull it back in.

> In summary, most of my concerns are that this will linger in our codebase beyond a few months, and that its actually not simpler and won't be fundamentally faster than just going back and doing this a less hack way. These are not huge concerns, so if this "gets the job done" and you're convinced that any less hack way is a huge timesink in comparison, I'd say go for it. But maybe its worth a quick re-evaluate if this hack is really as simple as it seemed at the outset?

Thanks.  Yeah I definitely don't have the cycles this sprint to do even a minimal DB backed version but it's also about the fact that I don't know that's the right approach and would like more data. This all started as a bonus I added and then reverted on [another PR](https://github.com/code-dot-org/code-dot-org/pull/67600) where it was clear we would need a tighter loop to iterate on prompts.  I think this solves that need in the immediate term (i.e. enabling the team's internal prompt development and testing - the fact that it could be shared with students for individual focus testing was more of a bonus-bonus) and gets us thinking about what to build and actually dedicate eng time to.

P.S. I was surprised by the 50 user GitHub limit but it looks like that's [just per 24 hours](https://docs.github.com/en/organizations/managing-membership-in-your-organization/inviting-users-to-join-your-organization#about-organization-invitations).

snickell reviewed (approved):
@edcodedotorg and I chatted, given the problem constraints, this seems like a reasonable hack to put in place for several months to get things moving (as long as we know we won't be doing it this way long-term).

breville commented:
This was a very interesting discussion!  Seth's concerns really resonated.  Some thoughts:
- I think this feature is a great idea, and can think of a lot of use cases for a self-serve prompt library across all of our AI efforts.
- I'm also very cautious about adding a new system, because we've seen small systems in the past take on a life of their own.  If this system becomes popular, it will be hard to transition users to anything that is less full-featured, which is a particular problem when trying to replicate what GitHub offers!  While we don't need to build for a decade of usage up front, this is an opportunity to at least make a plan for what we do if this system does get used for a decade.
- We clearly have a [missing system](https://codedotorg.slack.com/archives/C03CK49G9/p1742006437099049?thread_ts=1730271508.022859&cid=C03CK49G9), some kind of static storage solution that is useful for a variety of teams and purposes.  The closest we had was probably the Dropbox integration in Pegasus!
-  My suggestion for now is to actually use an S3 bucket.  We do that for the Music Lab library already.  Anyone with the right credential can use a client like [Cyberduck](https://cyberduck.io/) to manage content.  It won't have all the bells and whistles of some other solutions, but it's a supported part of our stack, and we'll still learn a lot from how it's used.  Implementation should be as simple as creating a new bucket and then retrieving files from the right public URL based on the provided URL parameter.

breville commented:
@edcodedotorg I've prepped a PR based on this one which uses S3.  Feel free to try it out and merge into here if it feels like a good fit.

edcodedotorg commented:
> @edcodedotorg As discussed, I've prepped a [PR](https://github.com/code-dot-org/code-dot-org/pull/68055) based on this one which uses S3. Feel free to try it out and merge into here if it feels like a good fit.

thanks and merged!  Will do some testing from here.

fisher-alice commented on apps/src/lab2/views/components/AiTutor2Chat.tsx:
```diff
@@ -18,14 +21,26 @@ import {aiTutorModelId} from '../../ai/ai-tutor-model-id';
   ? `You are an AI Computer Science Tutor that supports students through scaffolded learning, metacognitive reflection, and problem-solving strategies. Target the reading age of an American 7th grader. By default, when a student asks a question, you should respond with a clarifying question, a small hint, or a reflective nudge—to help them take the next step without solving the task for them. Do  ...
   : `You are an AI Computer Science Tutor that supports students through scaffolded learning, metacognitive reflection, and problem-solving strategies. Target the reading age of an American 7th grader. By default, when a student asks a question, you should respond with a clarifying question, a small hint, or a reflective nudge—to help them take the next step without solving the task for them. Do  ...
 
+// Optional name to retrieve custom system prompt from 'experimentation-settings' repo.
```
from s3?
